pkg/provider/passworddepot/passworddepot_api.go

```diff
@@ -0,0 +1,344 @@
+	databases, err := api.ListDatabases()
+	if err != nil {
+		return "", errors.Wrap(err, "error: getting database list")
+	}
+
+	for _, db := range databases.Databases {
+		if strings.Contains(db.Name, database) {
+			return db.Fingerprint, nil
+		}
+	}
+
+	return "", errDBNotFound
+}
+
+func (api *API) getSecretFingerprint(databaseFingerprint, secretName, folder string) (string, error) {
+	secrets, err := api.ListSecrets(databaseFingerprint, folder)
+	if err != nil {
+		return "", errors.Wrap(err, "error: getting secrets list")
+	}
+
+	parts := strings.Split(secretName, ".")
+	searchName := parts[0]
+	var fingerprint string
+	for _, entry := range secrets.Entries {
+		if strings.Contains(entry.Name, searchName) {
+			fingerprint = entry.Fingerprint
+			if len(parts) > 1 {
+				return api.getSecretFingerprint(databaseFingerprint, strings.Join(parts[1:], "."), fingerprint)
+			}
+			return fingerprint, nil
+		}
+	}
+	return "", errSecretNotFound
+}
+
+func (api *API) getendpointURL(endpoint string) string {
+	return fmt.Sprintf("https://%s:%s/v1.0/%s", api.baseURL, api.hostPort, endpoint)
+}
+
+func (api *API) login() error {
+	loginRequest, err := http.NewRequest("GET", api.getendpointURL("login"), http.NoBody)
+	if err != nil {
+		return errors.Wrap(err, "error creating request")
+	}
+	loginRequest.Header.Add("user", api.username)
+	loginRequest.Header.Add("pass", api.password)
+
+	resp, err := api.client.Do(loginRequest)
+	if err != nil {
+		return errors.Wrap(err, errorDoRequest)
+	}
+	defer func() {
+		if resp.Body != nil {
+			resp.Body.Close()
+		}
+	}()
+
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(resp.Body)
+	if err != nil {
+		return errors.Wrap(err, errorReadBody)
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		return fmt.Errorf("failed to authenticate with the given credentials: %d %s", resp.StatusCode, buf.String())
+	}
+
+	var accessData AccessData
+	err = json.Unmarshal(buf.Bytes(), &accessData)
+	if err != nil {
+		return errors.Wrap(err, "error: unmarshal body")
+	}
+
+	api.secret = &accessData
+
+	return nil
+}
+
+func (api *API) ListSecrets(dbFingerprint, folder string) (DatabaseEntries, error) {
+	endpointURL := api.getendpointURL(fmt.Sprintf("list?db=%s", dbFingerprint))
+	if folder != "" {
+		endpointURL = fmt.Sprintf("%s&folder=%s", endpointURL, folder)
+	}
+	listSecrets, err := http.NewRequest("GET", endpointURL, http.NoBody)
+	if err != nil {
+		return DatabaseEntries{}, errors.Wrap(err, "error: creating secrets request")
+	}
+
+	respSecretsList, err := api.doAuthenticatedRequest(listSecrets)
+	if err != nil {
+		return DatabaseEntries{}, errors.Wrap(err, errorDoRequest)
+	}
+	defer func() {
+		if respSecretsList.Body != nil {
+			respSecretsList.Body.Close()
+		}
+	}()
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(respSecretsList.Body)
+	if err != nil {
+		return DatabaseEntries{}, err
+	}
+	var dbEntries DatabaseEntries
+	err = json.Unmarshal(buf.Bytes(), &dbEntries)
+	return dbEntries, err
```

Comment:
This pattern of getting the body and unmarshalling checking error then returning repeats itself 4 times.

It warrants a refactor. You can do it by creating the unmarshal logic and passing in `any` and return the result from the function directly in the calling function.

Reply:
Nice! Since you are reading the whole body inside that function it's safe to shift the defer close inside that function too. So you can delete that whole defer thing in the other calls.

Further.. Since you only check the status code in login, if you shift the status check into that function, you'll get that check for free for the other times you are calling `Do(req)`. Doing that you can now delete the check from login. With all that, I think it's safe to pass in the response as is with the desired object to be marshalled to.

Reply:
Once you have that the code will be really clean and precise. :)